### PR TITLE
WiP: Simulating the Euler equations on multiple GPUs using Python

### PR DESCRIPTION
**Pull request by setmar**

### CODE REVIEW

File: GPUSimulators/SHMEMSimulator.py [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+        
+        #Shorthands for computing extents and sizes
+        gc_x = int(self.sim.getOutput()[0].x_halo)
+        gc_y = int(self.sim.getOutput()[0].y_halo)
+        nx = int(self.sim.nx)
+        ny = int(self.sim.ny)
+        
+        #Set regions for ghost cells to read from
+        #These have the format [x0, y0, width, height]
+        self.read_e = np.array([  nx,    0, gc_x, ny + 2*gc_y])
```
Comment (babrodtk) [inline GPUSimulators/SHMEMSimulator.py]: These should not be needed when we go for gpu-gpu direct copy. But for now it's best to use them and use the same setup as for MPI

File: GPUSimulators/SHMEMSimulator.py [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+    def synchronize(self):
+        self.sim.synchronize()
+        
+    def check(self):
+        return self.sim.check()
+    
+    def computeDt(self):
+        local_dt = np.array([np.float32(self.sim.computeDt())])
+        global_dt = np.empty(1, dtype=np.float32)
+        self.grid.comm.Allreduce(local_dt, global_dt, op=MPI.MIN)
```
Comment (babrodtk) [inline GPUSimulators/SHMEMSimulator.py]: MPI use here. Should be changed to a for loop to gather the minimum dt across the simulators

File: GPUSimulators/SHMEMSimulator.py [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+        x0 = i * width
+        y0 = j * height 
+        x1 = x0 + width
+        y1 = y0 + height
+        return [x0, x1, y0, y1]
+        
+    def exchange(self):        
+        ns_download_before_exchange()
+        # GLOBAL SYNC 
+        ns_do_exchange()
```
Comment (babrodtk) [inline GPUSimulators/SHMEMSimulator.py]: This should not be needed when not using MPI.

File: GPUSimulators/SHMEMSimulator.py [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+        #Download from the GPU
+        if self.north is not None:
+            for k in range(self.nvars):
+                self.sim.u0[k].download(self.sim.stream, cpu_data=self.out_n[k,:,:], asynch=True, extent=self.read_n)
+        if self.south is not None:
+            for k in range(self.nvars):
+                self.sim.u0[k].download(self.sim.stream, cpu_data=self.out_s[k,:,:], asynch=True, extent=self.read_s)
+        self.sim.stream.synchronize()
+
+    def ns_do_exchange(self):
```
Comment (babrodtk) [inline GPUSimulators/SHMEMSimulator.py]: This function should not be needed for shmem simulator

File: GPUSimulators/SHMEMSimulator.py [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+        width = self.sim.nx*self.sim.dx
+        height = self.sim.ny*self.sim.dy
+        i, j = self.grid.getCoordinate(self.index)
+        x0 = i * width
+        y0 = j * height 
+        x1 = x0 + width
+        y1 = y0 + height
+        return [x0, x1, y0, y1]
+        
+    def exchange(self):
```
Comment (babrodtk) [inline GPUSimulators/SHMEMSimulator.py]: This is perfect for version 1. For version 2 we should use direct gpu-gpu copy.

File: GPUSimulators/SHMEMSimulator.py [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+        #Download from the GPU
+        if self.east is not None:
+            for k in range(self.nvars):
+                self.sim.u0[k].download(self.sim.stream, cpu_data=self.out_e[k,:,:], asynch=True, extent=self.read_e)
+        if self.west is not None:
+            for k in range(self.nvars):
+                self.sim.u0[k].download(self.sim.stream, cpu_data=self.out_w[k,:,:], asynch=True, extent=self.read_w)
+        self.sim.stream.synchronize()
+
+    def ew_do_exchange(self):
```
Comment (babrodtk) [inline GPUSimulators/SHMEMSimulator.py]: Same as for ns - not needed

File: GPUSimulators/SHMEMSimulator.py [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+        #Upload to the GPU
+        if self.east is not None:
+            for k in range(self.nvars):
+                self.sim.u0[k].upload(self.sim.stream, self.in_e[k,:,:], extent=self.write_e)
+        if self.west is not None:
+            for k in range(self.nvars):
+                self.sim.u0[k].upload(self.sim.stream, self.in_w[k,:,:], extent=self.write_w)
+        
+        #Wait for sending to complete
+        for comm in comm_send:
```
Comment (babrodtk) [inline GPUSimulators/SHMEMSimulator.py]: Not needed for shmem

File: GPUSimulators/SHMEMSimulatorGroup.py [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+        self.ngpus = ngpus
+        self.ndims = ndims
+
+        self.grid = SHMEMGrid.getGrid(self.ngpus, self.ndims)
+        
+        self.logger.debug("Created {:}-dimensional SHMEM grid, using {:} GPUs".format(
+                self.ndims, self.ngpus))    
+
+        # XXX: Is this a natural place to store the contexts? Consider moving contexts out of this class.
+        self.cuda_contexts = []
```
Comment (babrodtk) [inline GPUSimulators/SHMEMSimulatorGroup.py:62]: Defined in simulator itself:
https://github.com/babrodtk/ShallowWaterGPU/blob/402e2c6f9f1efcc8cb4642be6f67015de78b686e/GPUSimulators/Simulator.py#L140

**Comment by babrodtk** [comment]
@setmar Is this outdated?

**Comment by setmar** [comment]
No. This is the paper (+something).

**Pull request merged**